### PR TITLE
J-UI-6: Per-blip task toggle + done state

### LINKED ISSUE #1084: J-UI-6: Per-blip task toggle + done state
## Parent

Umbrella: #1078
Tracker: #904

## User story

I click the task affordance on a blip; the blip becomes a task. I click "done"; the task is marked complete and the visual state reflects it (strikethrough/checkmark). Reload preserves the state.

## Acceptance

Task toggle is wired to the task DocOp. Done state renders with strikethrough/checkmark and persists across reload.

## Hard acceptance — matrix rows

- **R-5.4 (tasks and related metadata overlays)**: Task toggle, completion state, and associated metadata overlays preserved. State persists through live updates from other clients. Keyboard toggling preserved. Overlays announce state changes to AT. Labels translate.

## Implementation notes

Touch:
- `j2cl/lit/src/elements/wavy-task-affordance.js` — the per-blip toggle element.
- `j2cl/lit/src/elements/task-metadata-popover.js`, `j2cl/lit/src/elements/interaction-overlay-layer.js` — metadata overlays; confirm they don't trap focus.
- `j2cl/lit/src/elements/wave-blip.js`, `j2cl/lit/src/elements/wave-blip-toolbar.js` — host the affordance; ensure done state styling propagates.
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/overlay/J2clTaskItemModel.java` — task model.
- Wire the toggle to the existing task DocOp path; ensure the DocOp uses the StringBuilder pattern for any text-bearing fragments (project memory).
- Confirm done state is reflected in `J2clSelectedWaveModel` so live updates from other clients re-render correctly.

## Local-server verification

1. Boot local server; register a fresh user.
2. Open a wave with at least two blips. Depends on **J-UI-4** for blip rendering.
3. Click the task affordance on one blip; confirm it visually becomes a task.
4. Click "done"; confirm strikethrough/checkmark renders and the affordance state updates.
5. Reload the page; confirm both the task and done states persist.
6. Open the wave in a second browser session; toggle done off in session A — confirm session B re-renders without manual reload (live update).
7. Tab through the wave with the keyboard; confirm the task toggle is keyboard-reachable and Space/Enter toggles state with an AT-announced state change.
8. Screenshot the done-state blip and the post-reload state; attach to PR.

## Workflow

Full drill: plan → review → implement → review → QA on local server → flag → changelog → PR. PR gets a `wave-pr-monitor` pane that stays alive until merged with all conversations resolved.

## Out of scope

- Initial blip rendering / threading — owned by **J-UI-4** (and required as a dependency).
- Reactions overlays — covered by F-3 follow-ups #1073–#1076.
- Mark-as-read decrement — owned by **J-UI-7**.

### PR DESCRIPTION
**Pull request by vega113**
## Summary
- Plumb persisted `task/done` (and `task/assignee`, `task/dueTs`) through the J2CL read-surface model so the strikethrough/checkmark state survives reload **and** live updates from other clients in `?view=j2cl-root`. Pre-J-UI-6 the F-3.S2 toggle wrote the annotation correctly but the read surface never read it back, so the UI flickered on every re-render.
- Track in-flight self-toggles in the renderer (`noteOptimisticTaskState`) so an unrelated live update arriving inside the toggle's submit/echo window can no longer revert the optimistic strikethrough — Copilot review caught this race during the implementation review (commit `894f79681`).
- Hoist every visible / aria-label / aria-live string on `<wavy-task-affordance>` to lit properties so the Java view (or a future lit-i18n slice) can override per-locale without re-architecting the element. English defaults preserved verbatim.

## Closes
- Closes #1084
- Refs umbrella #1078, parent #904

## Matrix rows satisfied
- **R-5.4** (`docs/j2cl-gwt-parity-matrix.md:134`): "Task toggle, completion state, and associated metadata overlays preserved; state persists through live updates" — `J2clSelectedWaveProjector.documentTaskDone/Assignee/DueTimestamp` reads directly from the wavelet DocOp annotations (per project memory `feedback_search_no_conversation_model`); `J2clReadBlip` + `J2clReadWindowEntry` carry the state through both render paths; `J2clReadSurfaceDomRenderer.applyTaskState` surfaces it as `data-task-completed` / `data-task-assignee` / `data-task-due-date` on the rendered `<wave-blip>` element. "Keyboard toggling preserved" + "Overlays announce state changes" still satisfied via the F-3.S2 `role="checkbox"` + `aria-live="polite"` region. "Labels translate" addressed by hoisting strings to lit properties.

## Local-server verification
Browser-based screenshots were skipped because Chrome MCP was rate-limited mid-session. Text-only verification:
- `bash scripts/worktree-boot.sh --port 9900` → server up, `STATUS=200`.
- Registered fresh user `jui6test@local.net` via `/auth/register` (per `feedback_local_registration_before_login_testing` — never assume a user exists).
- Signed in via `/auth/signin` → 302 to `/`.
- `curl /?view=j2cl-root` returns the J2CL shell-root markup with `sidecar-selected-card` and the `/j2cl/assets/shell.js` bundle.
- `grep -c labelToggleOpen|labelAriaCheck|noteOptimisticTaskState war/j2cl/assets/shell.js` → 4 (new code is bundled).
- `grep -c data-task-completed|wave-blip-task-toggled war/j2cl/assets/shell.js` → 6.

Will attach screenshots via PR comment once Chrome MCP is back.

## Test plan
- [x] `cd j2cl && ./mvnw -Dgpg.skip=true test` — 800/800 passing (119 browser-only cases skipped on JVM).
- [x] `cd j2cl/lit && npm test` — 524/524 passing.
- [x] New Java cases on `J2clSelectedWaveProjectorTest` (8): `documentTaskDone` true/false, `documentTaskAssignee`, `documentTaskDueTimestamp` numeric/blank/garbage, `enrichReadBlipMetadata` propagation, `enrichWindowEntriesFromReadBlips` happy path / placeholders / empty inputs.
- [x] New Java cases on `J2clReadSurfaceDomRendererTest` (7): `data-task-completed` set/cleared on flat-render + window-render paths, done→open re-render clears the attribute, `formatDueDate` edge cases, optimistic-toggle override during in-flight, optimistic state clears when server echoes.
- [x] New Lit cases on `wavy-task-affordance.test.js` (1): label property override with French strings.
- [x] New Lit cases on `wave-blip.test.js` (2): strikethrough CSS lights up when `data-task-completed` is set, `taskAssignee` + `taskDueDate` propagate to the inner affordance.
- [x] Local server smoke: shell-root mounts, bundle loads.
- [ ] Browser-based reload + cross-client live-update screenshots — deferred to follow-up comment when Chrome MCP is back.

## Files touched
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadBlip.java` — `taskDone`, `taskAssignee`, `taskDueTimestamp`, `withTaskDone(boolean)`.
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadWindowEntry.java` — same fields + `loadedWithTaskMetadata(...)` factory + `withTaskDone(boolean)`.
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveProjector.java` — `documentTaskDone/Assignee/DueTimestamp` helpers; `enrichWindowEntriesFromReadBlips` bridge.
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java` — wires the enrichment bridge before `renderWindow`; binds the `wave-blip-task-toggled` listener to `contentList` for the optimistic-toggle hook.
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java` — `applyTaskState` writes/clears the three data-attributes, `noteOptimisticTaskState` registry, `sameReadBlip` + `sameWindowEntry` include task state in equality.
- `j2cl/lit/src/elements/wavy-task-affordance.js` — label properties.
- Tests + changelog fragment.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Per-blip task state (completed, assignee, due date) now persists through reloads and live updates
  * Task metadata is emitted as attributes so affordances and overlays show assignee and due date; completed tasks get visual strikethrough
  * Task affordance labels/ARIA/live text are configurable for localization
  * Optimistic UI toggles give immediate feedback and expire/clear if not confirmed

* **Tests**
  * Expanded unit and integration tests covering annotation parsing, rendering, optimistic behavior, and due-date formatting

* **Chores**
  * Added plan and changelog entry documenting the changes

### CODE REVIEW

**Comment by vega113** [comment]
## Merge Conflicts Resolved

The branch has been rebased onto the latest `main` and force-pushed. PR is now `MERGEABLE`.

### Conflicting Files
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java`
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveProjector.java`
- `j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java`

### Resolution Summary

**`J2clReadSurfaceDomRenderer.java` (`sameReadBlip`):**
Both sides added fields to the fast-path equality check: J-UI-4 (`parentBlipId`, `threadId` for manifest linkage) and J-UI-6 (`isTaskDone`, `getTaskAssignee`, `getTaskDueTimestamp` for task-state change detection). Both sets of fields were kept — they serve orthogonal purposes and must coexist.

**`J2clSelectedWaveProjector.java` and `J2clSelectedWaveView.java`:**
No textual conflict markers — git flagged both-modified but the working-tree content was already correct. Files staged to mark resolution and rebase continued.

### Verification
- [x] No conflict markers in any file
- [x] All J-UI-4 manifest linkage fields preserved
- [x] All J-UI-6 task-state fields preserved
- [x] Branch rebased cleanly onto `main`
- [x] `mergeable: MERGEABLE` confirmed via GitHub API

File: wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      "type": "feature",
+      "items": [
+        "R-5.4 reload + live-update persistence: opening a wave with a previously-completed task blip now paints the strikethrough/checkmark immediately (no flash of unchecked state), and toggling done from a second client re-renders the first client without manual reload.",
+        "Associated metadata overlays preserved (R-5.4): the task-metadata popover re-mounts with the persisted assignee + due-date when reopened on a reloaded blip, instead of starting from empty defaults.",
+        "Localizable labels (R-5.4 \"Labels translate\"): wavy-task-affordance hoists every visible / aria-label / aria-live string to a settable lit property (labelToggleOpen, labelToggleDone, labelAriaCheck, labelAriaUncheck, labelDetails, labelAnnounceDone, labelAnnounceOpen). English defaults preserved verbatim so the F-3.S2 baseline behavior is unchanged."
+      ]
+    },
+    {
+      "type": "internal",
+      "items": [
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json:18]: Fixed: the changelog already uses "feature" and "fix" section types in the current head commit (b0ebd8ea). The "internal" section was removed and items were folded into the existing "fix" section.

File: wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "releaseId": "2026-04-28-issue-1084-j-ui-6",
+  "version": "J-UI-6 (#1084)",
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json]: Fixed: the version field already reads "PR #1097" in the current head commit (b0ebd8ea) — normalized to the PR-number format used by other entries.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java [hunk 3]
```diff
@@ -262,6 +264,50 @@ private static HTMLElement ensureWaveNavRow(HTMLElement card) {
    * surface. S2 records telemetry for each click; S5 will add the
    * controller wiring on top.
    */
+  /**
+   * J-UI-6 (#1084, R-5.4): forward {@code wave-blip-task-toggled} events to
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java]: Fixed: the two Javadoc blocks are no longer adjacent in b0ebd8ea — bindOptimisticTaskToggleListener (lines 262-304) and bindChromeEvents (lines 306-311) each have their own attached Javadoc. No orphaned block remains.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveProjector.java [hunk 4]
```diff
@@ -588,6 +596,77 @@ static List<J2clReadBlip> applyConversationManifest(
...
+    boolean changed = false;
+    for (J2clReadWindowEntry entry : windowEntries) {
+      if (entry == null || !entry.isLoaded()) {
+        enriched.add(entry);
+        continue;
+      }
+      J2clReadBlip blip = blipsById.get(entry.getBlipId());
+      if (blip == null) {
+        enriched.add(entry);
+        continue;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveProjector.java:644]: Fixed: the catch-block comment in b0ebd8ea was updated to accurately describe the conservative reset behavior: "Unparseable values reset to 'unset' rather than rendering a corrupt epoch placeholder." — no longer claims to preserve the previous value.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 5]
```diff
@@ -900,6 +714,127 @@ private static void setProperty(HTMLElement element, String name, Object value)
...
+    Boolean optimistic = blipId == null ? null : optimisticTaskState.get(blipId);
+    if (optimistic != null) {
+      // Server caught up — clear the optimistic override and let the model
+      // win going forward. Otherwise apply the optimistic value over the
+      // (still-stale) model snapshot.
+      if (optimistic.booleanValue() == modelTaskDone) {
+        optimisticTaskState.remove(blipId);
+      } else {
+        taskDone = optimistic.booleanValue();
+      }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in b0ebd8ea: applyTaskState now uses a TTL-based OptimisticTaskEntry (expiresAtMs). Stale overrides past the deadline are dropped and model state wins, so a failed submit no longer leaves the override stuck indefinitely.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 6]
```diff
@@ -900,6 +714,127 @@ private static void setProperty(HTMLElement element, String name, Object value)
...
+      String blipId,
+      boolean modelTaskDone,
+      String taskAssignee,
+      long taskDueTimestamp) {
+    if (element == null) {
+      return;
+    }
+    boolean taskDone = modelTaskDone;
+    Boolean optimistic = blipId == null ? null : optimisticTaskState.get(blipId);
+    if (optimistic != null) {
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in b0ebd8ea: the optimistic state map is now keyed by optimisticTaskKey(currentWaveId(), blipId) — a compound (waveId, blipId) key — so blip ids cannot bleed across wave boundaries.

File: wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (vega113) [reply wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json:18]: Fixed in commit 68b1d96ca — changed the section type from `internal` to `fix`. The Server Build (JDK 17) validate-changelog step was indeed failing with the exact error you flagged.

File: wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (vega113) [reply wave/config/changelog.d/2026-04-28-issue-1084-j-ui-6.json]: Fixed in commit 68b1d96ca — switched the version field to `PR #1097` to match the surrounding J-UI fragments.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveView.java]: Fixed in commit 68b1d96ca — moved the orphan javadoc back to `bindChromeEvents` where it originated.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveProjector.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSelectedWaveProjector.java:644]: Fixed in commit 68b1d96ca — updated the comment to describe the actual conservative behaviour (corrupt range overrides earlier successful parse).

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in commit 68b1d96ca — added a 30-second TTL on each optimistic entry. `applyTaskState` drops the entry once `expiresAtMs <= currentTimeMs()`, so a failed submit no longer pins the override forever. Test coverage added: `optimisticTaskStateExpiresAfterTtlOnFailedSubmit`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in commit 68b1d96ca — composite key is now `waveId + U+0001 + blipId`. The view's body listener reads waveId from the event detail, falling back to the content-list host's `data-wave-id` attribute. The renderer reads its own host's attribute on each apply. Test coverage added: `optimisticTaskStateIsNamespacedByWaveId` exercises the wave-A → wave-B no-bleed scenario.

File: docs/superpowers/plans/2026-04-28-j-ui-6-plan.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+### Verification commands (before opening PR — required by user feedback)
+
+```
+cd wave
+sbt compile
+sbt "j2cl/Test/test"
+cd ../j2cl/lit && npm test
+cd ../.. && sbt "; project root; test:compile"
+```
```
[the remark this answers is not in the document]
Comment (vega113) [reply docs/superpowers/plans/2026-04-28-j-ui-6-plan.md]: Fixed in commit 096cf47d9 — added `sh` language tag to the verification-commands fence.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java [hunk 8]
```diff
@@ -900,6 +714,127 @@ private static void setProperty(HTMLElement element, String name, Object value)
...
+   * <p>Idempotent: a second call with the same value is a no-op. Setting
+   * the state explicitly to the model's already-projected value clears
+   * the entry (the optimistic intent is satisfied).
+   */
+  public void noteOptimisticTaskState(String blipId, boolean completed) {
+    if (blipId == null || blipId.isEmpty()) {
+      return;
+    }
+    optimisticTaskState.put(blipId, Boolean.valueOf(completed));
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/read/J2clReadSurfaceDomRenderer.java]: Fixed in commit 68b1d96ca — `noteOptimisticTaskState` now wraps each entry in `OptimisticTaskEntry(done, expiresAtMs)` with a 30-second TTL (`OPTIMISTIC_TASK_TTL_MS = 30_000L`). `applyTaskState` drops the entry and falls back to model state once `expiresAtMs <= currentTimeMs()`, so a failed submit no longer pins the override indefinitely. Test coverage added: `optimisticTaskStateExpiresAfterTtlOnFailedSubmit`.
